components/brave_news/browser/initialization_promise.h

```diff
@@ -0,0 +1,66 @@
+// Copyright (c) 2024 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#ifndef BRAVE_COMPONENTS_BRAVE_NEWS_BROWSER_INITIALIZATION_PROMISE_H_
+#define BRAVE_COMPONENTS_BRAVE_NEWS_BROWSER_INITIALIZATION_PROMISE_H_
+
+#include <string>
+
+#include "base/functional/callback_forward.h"
+#include "base/memory/raw_ref.h"
+#include "base/memory/weak_ptr.h"
+#include "base/one_shot_event.h"
+#include "brave/components/brave_news/browser/publishers_controller.h"
+
+namespace brave_news {
+
+class BraveNewsPrefManager;
+class PublishersController;
+
+class InitializationPromise {
+ public:
+  enum class State { kNone, kInitializing, kInitialized, kFailed };
+
+  InitializationPromise(size_t max_retries,
+                        BraveNewsPrefManager& pref_manager,
+                        PublishersController& publishers_controller);
+  ~InitializationPromise();
+
+  void OnceInitialized(base::OnceClosure on_initialized);
+
+  State state() { return state_; }
+  bool failed() { return state_ == InitializationPromise::State::kFailed; }
+  bool complete() { return on_initializing_prefs_complete_.is_signaled(); }
+
+  void set_no_retry_delay_for_testing(bool no_retry_delay) {
+    no_retry_delay_for_testing_ = no_retry_delay;
+  }
+
+  size_t attempts_for_testing() { return attempts_; }
+
+ private:
+  void Initialize();
+  void OnGotLocale(const std::string& locale);
+
+  void Notify();
+
+  raw_ref<BraveNewsPrefManager> pref_manager_;
+  raw_ref<PublishersController> publishers_controller_;
+
+  base::OneShotEvent on_initializing_prefs_complete_;
+
+  bool no_retry_delay_for_testing_ = false;
+
+  size_t max_retries_ = 0;
```

Comment:
optional: I think this can be mark const as it's assigned in the constructor and unchanged

```suggestion
  const size_t max_retries_ = 0;
```